c++/src/kj/compat/gzip-test.c++

```diff
@@ -0,0 +1,174 @@
+    KJ_EXPECT(gzip.readAllText().wait(io.waitScope) == "foobar");
+  }
+
+  // Force read one byte at a time.
+  {
+    MockInputStream rawInput(FOOBAR_GZIP, 1);
+    GzipInputStream gzip(rawInput);
+    KJ_EXPECT(gzip.readAllText().wait(io.waitScope) == "foobar");
+  }
+
+  // Read truncated input.
+  {
+    MockInputStream rawInput(kj::arrayPtr(FOOBAR_GZIP, sizeof(FOOBAR_GZIP) / 2), kj::maxValue);
+    GzipInputStream gzip(rawInput);
+
+    char text[16];
+    size_t n = gzip.tryRead(text, 1, sizeof(text)).wait(io.waitScope);
+    text[n] = '\0';
+    KJ_EXPECT(StringPtr(text, n) == "fo");
+
+    KJ_EXPECT_THROW_MESSAGE("gzip compressed stream ended prematurely",
+        gzip.tryRead(text, 1, sizeof(text)).wait(io.waitScope));
+  }
+
+  // Read concatenated input.
+  {
+    Vector<byte> bytes;
+    bytes.addAll(ArrayPtr<const byte>(FOOBAR_GZIP));
+    bytes.addAll(ArrayPtr<const byte>(FOOBAR_GZIP));
+    MockInputStream rawInput(bytes, kj::maxValue);
+    GzipInputStream gzip(rawInput);
+
+    KJ_EXPECT(gzip.readAllText().wait(io.waitScope) == "foobarfoobar");
+  }
+}
+
+class MockOutputStream: public AsyncOutputStream {
+public:
+  kj::Vector<byte> bytes;
+
+  kj::String decompress(WaitScope& ws) {
+    MockInputStream rawInput(bytes, kj::maxValue);
+    GzipInputStream gzip(rawInput);
+    return gzip.readAllText().wait(ws);
+  }
+
+  Promise<void> write(const void* buffer, size_t size) override {
+    bytes.addAll(arrayPtr(reinterpret_cast<const byte*>(buffer), size));
+    return kj::READY_NOW;
+  }
+  Promise<void> write(ArrayPtr<const ArrayPtr<const byte>> pieces) override {
+    for (auto& piece: pieces) {
+      bytes.addAll(piece);
+    }
+    return kj::READY_NOW;
+  }
+};
+
+KJ_TEST("gzip compression") {
+  auto io = setupAsyncIo();
+
+  // Normal write.
+  {
+    MockOutputStream rawOutput;
+    GzipOutputStream gzip(rawOutput);
+    gzip.write("foobar", 6).wait(io.waitScope);
+    gzip.end().wait(io.waitScope);
+
+    KJ_EXPECT(rawOutput.decompress(io.waitScope) == "foobar");
+  }
+
+  // Multi-part write.
+  {
+    MockOutputStream rawOutput;
+    GzipOutputStream gzip(rawOutput);
+    gzip.write("foo", 3).wait(io.waitScope);
+    gzip.write("bar", 3).wait(io.waitScope);
+    gzip.end().wait(io.waitScope);
+
+    KJ_EXPECT(rawOutput.decompress(io.waitScope) == "foobar");
+  }
+}
+
+KJ_TEST("gzip huge round trip") {
+  auto io = setupAsyncIo();
+
+  auto bytes = heapArray<byte>(65536);
+  for (auto& b: bytes) {
+    b = rand();
+  }
+
+  MockOutputStream rawOutput;
+  GzipOutputStream gzipOut(rawOutput);
+  gzipOut.write(bytes.begin(), bytes.size()).wait(io.waitScope);
+  gzipOut.end().wait(io.waitScope);
+
+  MockInputStream rawInput(rawOutput.bytes, kj::maxValue);
+  GzipInputStream gzipIn(rawInput);
+  auto decompressed = gzipIn.readAllBytes().wait(io.waitScope);
+
+  KJ_ASSERT(decompressed.size() == bytes.size());
+  KJ_ASSERT(memcmp(bytes.begin(), decompressed.begin(), bytes.size()) == 0);
+}
+
```

Comment:
Maybe add a test, or a TODO, for the array-of-arrays overload of `write()`?

Reply:
added test